amplify/backend/api/colonycdapp/schema.graphql

```diff
@@ -501,4 +513,99 @@ type ColonyAction @model {
+  #
+  # This should be a list/array, but this is yet another thing Amplify is bad at
+  # The official response from them is to use DynamoDB directly:
+  # https://stackoverflow.com/questions/50992106/append-item-to-list-using-aws-appsync-to-dynamodb
+  # Which is just freaking great!
+  #
+  # As such, I've decided to simplify this, an just pass a stringified JSON around
+  # since this is for display purpouses only.
+  #
+  # If anyone has a better idea on how to handle this, I'll gladly change it!
+  #
+  # When / if you decide to use it, just be mindful of what format other people
+  # are using, and follow suit.
+  #
+  # Roles Types:
+  # [{ id: String, type: String, role: Number, setTo: boolean }]
+  individualEvents: String
+}
+
+type ColonyActionRoles {
+  role_0: Boolean # recovery
+  role_1: Boolean # root
+  role_2: Boolean # arbitration
+  role_3: Boolean # architecture
+  role_5: Boolean # funding
+  role_6: Boolean # administration
+}
+
+# Roles have been set as role_<number> so we can more easily map them to their
+# actual contract counterpart on the client and block ingestor side(s)
+
+type ColonyRole @model {
+  id: ID! @index # colonyAddress_domainNativeId_userAddress_roles
+  domain: Domain! @hasOne
+
+  targetAddress: ID # Amplify will automatically populate one of the following fields with related model if it finds one
+  targetUser: User @hasOne(fields: ["targetAddress"])
+  targetColony: Colony @hasOne(fields: ["targetAddress"])
+  targetExtension: ColonyExtension @hasOne(fields: ["targetAddress"])
+  targetToken: Token @hasOne(fields: ["targetAddress"])
+
+  latestBlock: Int!
+  role_0: Boolean # recovery
+  role_1: Boolean # root
+  role_2: Boolean # arbitration
+  role_3: Boolean # architecture
+  role_5: Boolean # funding
+  role_6: Boolean # administration
+}
+
+# it's a model, not a type since this is designed to work outside of the main
+# colony model, or colony role model, as, by design it should be only used
+# in very specific cases
+#
+# This should take a snapshot of the user's full roles/permissions at a specific block
+
+type ColonyHistoricRole @model {
+  id: ID! # colonyAddress_domainNativeId_userAddress_blockNumber_roles
+
+  # I HATE AMPLIFY SORTING! MAY IT DIE A THOUSAND DEATHS!
+  #
+  # @NOTE Always set to `type` to "SortedHistoricRole" to enable Amplify sorting, it DOES NOT WORK with `id`
+  #
+  # Apparently, it can only filter non-primary fields, meaning that if we sort by `blockNumber` then that
+  # field is considered primary, so it won't allow us to filter by it.
+  # To overcome this we've sorted by creation date, which is not really correct from a block chain
+  # point of view, but due to the nature of the block ingestor, it works for our use case
+  type: String!
+    @index(
+      name: "getColonyHistoricRoleByDate"
+      queryField: "getColonyHistoricRoleByDate"
+      sortKeyFields: ["createdAt"]
+    )
+
+  domain: Domain! @hasOne
+  colony: Colony! @hasOne
```

Comment:
Nitpick, but I like to manually define the foreign key fields for these (e.g. add `colonyId` of type `ID` and pass it to the `@hasOne` decorator), as otherwise the generated names just end up a bit confusing:
<img width="286" alt="image" src="https://user-images.githubusercontent.com/112586815/234130036-61468c2d-88bd-4ce6-8a3c-7492af581e6d.png">

Reply:
Good idea, didn't think about it.